### PR TITLE
[Bugfix][Qwen3-TTS] Handle codec generations that exhaust their token budget

### PR DESCRIPTION
**Pull request by Sy0307**
### Context

Qwen3-TTS Base can occasionally fail to emit codec EOS and keep generating until `max_new_tokens`. The same symptom appears in vLLM-Omni #5889, #4815, and #6158, and in the official Qwen repository:

- [Qwen3-TTS #118](https://github.com/QwenLM/Qwen3-TTS/issues/118): approximately 1 in 200 Base generations reached `max_new_tokens` without EOS under the official Transformers path.
- [Qwen3-TTS #318](https://github.com/QwenLM/Qwen3-TTS/issues/318): mixed Thai/Chinese/English input made the failure substantially easier to reproduce.
- [Qwen3-TTS discussion #211](https://github.com/QwenLM/Qwen3-TTS/discussions/211): the issue was reported with Base and fine-tuned checkpoints and with both FlashAttention 2 and SDPA.

### Root-cause investigation

The investigation separated the model termination failure from the framework behavior that returned the resulting audio as success.

#### 1. Stop-token configuration parity

We compared the official Qwen generation code, checkpoint config, and the vLLM-Omni stage-0 sampling parameters. All three use codec EOS token `2150`; vLLM-Omni correctly passes it as `stop_token_ids=[2150]`. The problem is therefore not a missing or incorrectly mapped stop-token configuration.

#### 2. Official Transformers reproduction without vLLM

We ran the official single-request Transformers implementation with `Qwen/Qwen3-TTS-12Hz-1.7B-Base`, using the sampling settings from the upstream report:

```text
temperature=0.9
top_k=50
top_p=1.0
repetition_penalty=1.05
```

The seed scan used 800 independent seeds and the prompt:

```text
而奥运藏品作为弘扬奥运精神，普及奥运文化的重要载体。
```

Results:

- seed `100423` reached `max_new_tokens=192`, produced 191 decoded codec frames, and did not emit EOS;
- increasing the same request to `max_new_tokens=1024` produced 1023 frames and still did not emit EOS;
- the tail entered a repetitive codec-token loop rather than approaching a delayed EOS;
- this reproduced the failure with one request, no vLLM scheduler, no request batch, and no residual MTP batch sampling.

This rules out batch invariance as the root cause. Making residual MTP sampling scalar may change a stochastic trajectory, but it cannot eliminate a failure that already occurs in the official single-request implementation.

#### 3. EOS logit/rank tracing

We traced the EOS probability before top-k filtering for a normal seed and the failing seed:

- normal seed: EOS reached rank 13 at step 74, rank 7 at step 76, then rank 1 with probability `0.99226` at step 77; generation stopped normally;
- failing seed: through at least step 320, EOS was generally around ranks 100-1400 with probability around `1e-8` to `1e-7`;
- with `top_k=50`, the failing seed's EOS candidate was removed from the sampling set, so it could not be selected even though the token ID and stopping configuration were correct.

The direct cause is a model/checkpoint trajectory in which codec EOS becomes extremely unlikely while the codec tokens enter a repetitive loop.

#### 4. Eager and graph comparison

Eager execution can produce a different sampled trajectory and may make a particular request appear fixed. However:

- the failure is stochastic and seed-dependent;
- the official Transformers path reproduces it without vLLM CUDA graphs;
- the same class of failure is reported upstream with multiple attention backends.

There is therefore no evidence that disabling CUDA graphs is a root fix. This PR does not enable `enforce_eager`, change model logits, or replace the batch sampler with per-request scalar sampling.

#### 5. Framework failure identified

Before this change, stage 0's terminal `finish_reason` was not preserved in the final speech response path. Consequently, a codec generation ending because of `length` was decoded and returned as successful audio, potentially after running to the large configured limit.

The model can still enter the bad trajectory. The framework can reliably fix the surrounding behavior: apply a bounded default budget, detect the real `length` finish, discard incomplete audio, retry only when doing so preserves the request contract, and expose an explicit terminal error otherwise.

Issue interpretation:

- #5889 contains the model EOS failure plus the framework symptom of returning/decoding the runaway generation as success;
- #4815 shows that eager mode can alter the sampled path, but does not establish a CUDA-graph correctness bug;
- #6158 requests the runtime safeguard implemented by this PR.

### Behavior

This PR:

- propagates the real stage-0 `finish_reason` through request metrics and treats only `finish_reason == "length"` as budget exhaustion (there is no output-token-count heuristic);
- adds a Qwen3-TTS adapter hook for completed-generation validation instead of adding another model branch to `serving_speech.py`;
- applies a conservative Base default ceiling of `min(configured_limit, max(192, 12 * text_tokens))`;
- falls back to the configured limit if text tokenization is unavailable instead of incorrectly truncating at 192 frames;
- preserves an explicit `max_new_tokens` exactly as the caller's opt-out from the automatic ceiling;
- discards incomplete non-streaming audio and retries once with a fresh seed only when neither `seed` nor `max_new_tokens` was explicitly supplied;
- returns HTTP 500 after an explicit-control failure or a failed retry;
- emits a terminal structured error for SSE and WebSocket streams, including `partial_audio: true` and `action: "discard"` only when audio has already been emitted; raw audio streaming terminates the connection instead of ending cleanly;
- documents the behavior in the Qwen3-TTS recipe.

The normal path adds no model forward pass, sampler change, synchronization, or deep copy. It performs a shallow copy of the stage sampling parameters, one cached tokenizer call, and a terminal metrics check. A second generation occurs only on the rare non-streaming failure path.

Related: #5889, #4815, #6158.

## Test Plan

**vLLM Version:** `0.28.0`

**vLLM-Omni Commit:** `6ffdf1a8f1e47b01159b73b4cdd4c037e757dbba`

### Automated tests

Run on an 8x NVIDIA H200 host (Python 3.12.3):

```bash
PYTHONPATH=$PWD python -m pytest -q -o addopts= \
  tests/engine/test_orchestrator.py \
  tests/entrypoints/openai_api/test_serving_speech_stream.py \
  tests/entrypoints/openai_api/test_speech_usage.py \
  tests/entrypoints/openai_api/test_tts_adapter.py
```

Result: `132 passed`.

```bash
PYTHONPATH=$PWD python -m pytest -q -o addopts= \
  tests/entrypoints/openai_api/test_serving_speech.py \
  -k "raw_stream_passes_tts_params_to_common_guard or qwen3_tts_nonstream_retries_codec_limit_once or qwen3_tts_nonstream_does_not_retry_explicit_sampling_controls or nonstream_does_not_retry_nonretryable_generation_error or generate_audio_chunks_rejects_qwen3_codec_limit_before_success or sse_codec_limit_marks_partial_audio_for_discard or sse_error_before_audio_is_not_marked_partial"
```

Result: `8 passed, 238 deselected`.

```bash
git diff --name-only origin/main...HEAD -z | \
  xargs -0 env SKIP=mypy-3.10 uvx pre-commit run --files
```

Result: all applicable hooks passed, including Ruff, markdownlint, test markers, SPDX, and the TTS adapter migration ratchet. The repository currently skips `mypy-3.10` in CI because the all-files baseline is not clean; the changed-file run reports existing errors in the touched legacy modules, and no newly introduced mypy error remains.

## Test Result

H200 end-to-end checks with `Qwen3-TTS-12Hz-1.7B-Base` covered both normal serving and failure handling:

- normal Chinese request: HTTP 200, 80 codec tokens, valid 24 kHz WAV, 303,404 response bytes, 6.32 seconds of audio;
- explicit `max_new_tokens=2`: HTTP 500 with a codec-budget error instead of returning incomplete audio;
- mixed-language trigger based on Qwen3-TTS #318: the first generation reached the automatic 360-frame ceiling, the server discarded it and retried with fresh seed `2968015596`, the retry stopped normally at 69 codec tokens, and the request returned HTTP 200;
- raw streaming, SSE, and WebSocket tests verify that a length-finished generation cannot report clean success and that clients are told to discard partial audio where a structured terminal event is available.

### CODE REVIEW

**Review by linyueqian — APPROVED** [review]
lgtm
